src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs

```diff
@@ -573,9 +593,33 @@ private void ProcessEvents()
                                 watcher.NotifyFileSystemEventArgs(WatcherChangeTypes.Changed, expandedName);
                                 break;
                             case Interop.libc.NotifyEvents.IN_MOVED_FROM:
-                                previousEventName = expandedName;
-                                previousEventParent = isDir ? associatedDirectoryEntry : null;
-                                previousEventCookie = nextEvent.cookie;
+                                // We need to check if this MOVED_FROM event is standalone - meaning the item was moved out
+                                // of scope. We do this by checking if we are at the end of our buffer (meaning no more events) 
+                                // and if there is data to be read by polling the fd. If there aren't any more events, fire the
+                                // deleted event; if there are more events, handle it via next pass.
+                                if (_bufferPos == _bufferAvailable)
+                                {
+                                    Interop.libc.PollFlags resultFlags;
+                                    int result = Interop.libc.poll(_inotifyHandle.DangerousGetHandle().ToInt32(), Interop.libc.PollFlags.POLLIN, 0, out resultFlags);
+                                    if ((result == 1) && ((resultFlags & Interop.libc.PollFlags.POLLIN) == Interop.libc.PollFlags.POLLIN))
+                                    {
+                                        // more data is coming so cache for the next read
+                                        previousEventName = expandedName;
+                                        previousEventParent = isDir ? associatedDirectoryEntry : null;
+                                        previousEventCookie = nextEvent.cookie;
+                                    }
+                                    else
+                                    {
+                                        // There isn't any more data in the queue so this is a deleted event
+                                        watcher.NotifyFileSystemEventArgs(WatcherChangeTypes.Deleted, expandedName);
+                                    }
+                                }
+                                else
+                                {
+                                    previousEventName = expandedName;
+                                    previousEventParent = isDir ? associatedDirectoryEntry : null;
+                                    previousEventCookie = nextEvent.cookie;
```

Comment:
This code is now duplicated in two places.  Can we avoid that? e.g. pseudo-code
```
if (AtEndOfBuffer() && !EventWaitingOnDescriptor())
{
    RaiseDeletedEvent();
}
else
{
    previousEvent* = ...;
}
```
?